From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Theo Paris <theo@theoparis.com>
Date: Tue, 15 Sep 2026 20:21:39 -0700
Subject: [PATCH 03/11] build: add OSS hardware board config and no-KDK build
 fixes

  * doconf gains -osshw / MASTER_CONFIG_OSS_HARDWARE, threaded through
    every conf/Makefile, so OSS-hardware boards (qemu, superbird,
    ipad41) select the non-Apple-silicon config paths.
  * MakeInc.def registers the QEMU/SUPERBIRD/IPAD41 machine configs and
    flags, OSSHW_SOCS, the arm64 ARCH_STRING fallback, and the
    KernelSupport archive/link handling for those boards.
  * make_symbol_aliasing.sh resolves availability.pl via FAKEROOT_DIR
    instead of DRIVERKITROOT, and Libsyscall.xcconfig no longer includes
    a DEVELOPER_DIR-relative BSD.xcconfig, so a bare SDK layout builds
    without a KDK.
---
 SETUP/config/doconf             |  5 ++++-
 bsd/conf/Makefile               |  2 +-
 bsd/sys/make_symbol_aliasing.sh |  2 +-
 iokit/conf/Makefile             |  2 +-
 libkern/conf/Makefile           |  2 +-
 libsa/conf/Makefile             |  2 +-
 libsyscall/Libsyscall.xcconfig  |  2 +-
 makedefs/MakeInc.def            | 18 ++++++++++++++++--
 osfmk/conf/Makefile             |  2 +-
 pexpert/conf/Makefile           |  2 +-
 san/conf/Makefile               |  2 +-
 security/conf/Makefile          |  2 +-
 12 files changed, 30 insertions(+), 13 deletions(-)

diff --git a/SETUP/config/doconf b/SETUP/config/doconf
index 296699b3..36e88ba5 100755
--- a/SETUP/config/doconf
+++ b/SETUP/config/doconf
@@ -163,6 +163,9 @@ while ($#argv >= 1)
 	case "-tag":
 	    set DEF_TAG="-DMASTER_CONFIG_ENABLE_KERNEL_TAG"
 	    breaksw
+	case "-osshw":
+	    set DEF_OSSHW="-DMASTER_CONFIG_OSS_HARDWARE"
+	    breaksw
 	default:
 	    echo "${prog}: ${argv[1]}: unknown switch"
 	    exit 1
@@ -209,7 +212,7 @@ foreach SYS ($argv)
     cat $MASTER $MASTER_CPU - \
         $MASTER $MASTER_CPU \
     | \
-    unifdef -t -DPLATFORM_${PLATFORM} -DCPU_$cpu -DSOC_CONFIG_${SOC_CONFIG} -DSYS_${SYS} $DEF_AVP $DEF_SEED $DEF_EXCLAVES $DEF_SPTM $DEF_BUILD $DEF_TAG - \
+    unifdef -t -DPLATFORM_${PLATFORM} -DCPU_$cpu -DSOC_CONFIG_${SOC_CONFIG} -DSYS_${SYS} $DEF_AVP $DEF_SEED $DEF_EXCLAVES $DEF_SPTM $DEF_BUILD $DEF_TAG $DEF_OSSHW - \
     | \
     sed -n \
 	-e "/^+/{" \
diff --git a/bsd/conf/Makefile b/bsd/conf/Makefile
index d6dceb52..dc33bfd9 100644
--- a/bsd/conf/Makefile
+++ b/bsd/conf/Makefile
@@ -23,7 +23,7 @@ endif
 
 $(TARGET)/$(CURRENT_KERNEL_CONFIG)/Makefile: $(SRCROOT)/SETUP/config/doconf $(OBJROOT)/SETUP/config $(DOCONFDEPS)
 	$(_v)$(MKDIR) $(TARGET)/$(CURRENT_KERNEL_CONFIG)
-	$(_v)$(SRCROOT)/SETUP/config/doconf -c -cpu $(DOCONF_ARCH_CONFIG_LC) -soc $(CURRENT_MACHINE_CONFIG_LC) -platform $(PLATFORM_FOR_CONFIG) -d $(TARGET)/$(CURRENT_KERNEL_CONFIG) -s $(SOURCE) -m $(MASTERCONFDIR) $(DOCONF_SEED) $(DOCONF_EXCLAVES) $(DOCONF_SPTM) $(DOCONF_KERNEL_TAG) $(CURRENT_KERNEL_CONFIG)
+	$(_v)$(SRCROOT)/SETUP/config/doconf -c -cpu $(DOCONF_ARCH_CONFIG_LC) -soc $(CURRENT_MACHINE_CONFIG_LC) -platform $(PLATFORM_FOR_CONFIG) -d $(TARGET)/$(CURRENT_KERNEL_CONFIG) -s $(SOURCE) -m $(MASTERCONFDIR) $(DOCONF_SEED) $(DOCONF_EXCLAVES) $(DOCONF_SPTM) $(DOCONF_KERNEL_TAG) $(DOCONF_OSSHW) $(CURRENT_KERNEL_CONFIG)
 
 do_all: $(TARGET)/$(CURRENT_KERNEL_CONFIG)/Makefile
 	$(_v)${MAKE} \
diff --git a/bsd/sys/make_symbol_aliasing.sh b/bsd/sys/make_symbol_aliasing.sh
index 24fb244d..4c9f4789 100755
--- a/bsd/sys/make_symbol_aliasing.sh
+++ b/bsd/sys/make_symbol_aliasing.sh
@@ -34,7 +34,7 @@ fi
 SDKROOT="$1"
 OUTPUT="$2"
 
-AVAILABILITY_PL="${SDKROOT}/${DRIVERKITROOT}/usr/local/libexec/availability.pl"
+AVAILABILITY_PL="${FAKEROOT_DIR}/usr/local/libexec/availability.pl"
 
 if [ ! -x "${AVAILABILITY_PL}" ] ; then
     echo "Unable to locate ${AVAILABILITY_PL} (or not executable)" >&2
diff --git a/iokit/conf/Makefile b/iokit/conf/Makefile
index 61146d56..7ded8cc7 100644
--- a/iokit/conf/Makefile
+++ b/iokit/conf/Makefile
@@ -22,7 +22,7 @@ endif
 
 $(TARGET)/$(CURRENT_KERNEL_CONFIG)/Makefile: $(SRCROOT)/SETUP/config/doconf $(OBJROOT)/SETUP/config $(DOCONFDEPS)
 	$(_v)$(MKDIR) $(TARGET)/$(CURRENT_KERNEL_CONFIG)
-	$(_v)$(SRCROOT)/SETUP/config/doconf -c -cpu $(DOCONF_ARCH_CONFIG_LC) -soc $(CURRENT_MACHINE_CONFIG_LC) -platform $(PLATFORM_FOR_CONFIG) -d $(TARGET)/$(CURRENT_KERNEL_CONFIG) -s $(SOURCE) -m $(MASTERCONFDIR) $(DOCONF_SEED) $(DOCONF_EXCLAVES) $(DOCONF_SPTM) $(DOCONF_KERNEL_TAG) $(CURRENT_KERNEL_CONFIG)
+	$(_v)$(SRCROOT)/SETUP/config/doconf -c -cpu $(DOCONF_ARCH_CONFIG_LC) -soc $(CURRENT_MACHINE_CONFIG_LC) -platform $(PLATFORM_FOR_CONFIG) -d $(TARGET)/$(CURRENT_KERNEL_CONFIG) -s $(SOURCE) -m $(MASTERCONFDIR) $(DOCONF_SEED) $(DOCONF_EXCLAVES) $(DOCONF_SPTM) $(DOCONF_KERNEL_TAG) $(DOCONF_OSSHW) $(CURRENT_KERNEL_CONFIG)
 
 do_all: $(TARGET)/$(CURRENT_KERNEL_CONFIG)/Makefile
 	$(_v)${MAKE} \
diff --git a/libkern/conf/Makefile b/libkern/conf/Makefile
index d6dceb52..dc33bfd9 100644
--- a/libkern/conf/Makefile
+++ b/libkern/conf/Makefile
@@ -23,7 +23,7 @@ endif
 
 $(TARGET)/$(CURRENT_KERNEL_CONFIG)/Makefile: $(SRCROOT)/SETUP/config/doconf $(OBJROOT)/SETUP/config $(DOCONFDEPS)
 	$(_v)$(MKDIR) $(TARGET)/$(CURRENT_KERNEL_CONFIG)
-	$(_v)$(SRCROOT)/SETUP/config/doconf -c -cpu $(DOCONF_ARCH_CONFIG_LC) -soc $(CURRENT_MACHINE_CONFIG_LC) -platform $(PLATFORM_FOR_CONFIG) -d $(TARGET)/$(CURRENT_KERNEL_CONFIG) -s $(SOURCE) -m $(MASTERCONFDIR) $(DOCONF_SEED) $(DOCONF_EXCLAVES) $(DOCONF_SPTM) $(DOCONF_KERNEL_TAG) $(CURRENT_KERNEL_CONFIG)
+	$(_v)$(SRCROOT)/SETUP/config/doconf -c -cpu $(DOCONF_ARCH_CONFIG_LC) -soc $(CURRENT_MACHINE_CONFIG_LC) -platform $(PLATFORM_FOR_CONFIG) -d $(TARGET)/$(CURRENT_KERNEL_CONFIG) -s $(SOURCE) -m $(MASTERCONFDIR) $(DOCONF_SEED) $(DOCONF_EXCLAVES) $(DOCONF_SPTM) $(DOCONF_KERNEL_TAG) $(DOCONF_OSSHW) $(CURRENT_KERNEL_CONFIG)
 
 do_all: $(TARGET)/$(CURRENT_KERNEL_CONFIG)/Makefile
 	$(_v)${MAKE} \
diff --git a/libsa/conf/Makefile b/libsa/conf/Makefile
index d6dceb52..dc33bfd9 100644
--- a/libsa/conf/Makefile
+++ b/libsa/conf/Makefile
@@ -23,7 +23,7 @@ endif
 
 $(TARGET)/$(CURRENT_KERNEL_CONFIG)/Makefile: $(SRCROOT)/SETUP/config/doconf $(OBJROOT)/SETUP/config $(DOCONFDEPS)
 	$(_v)$(MKDIR) $(TARGET)/$(CURRENT_KERNEL_CONFIG)
-	$(_v)$(SRCROOT)/SETUP/config/doconf -c -cpu $(DOCONF_ARCH_CONFIG_LC) -soc $(CURRENT_MACHINE_CONFIG_LC) -platform $(PLATFORM_FOR_CONFIG) -d $(TARGET)/$(CURRENT_KERNEL_CONFIG) -s $(SOURCE) -m $(MASTERCONFDIR) $(DOCONF_SEED) $(DOCONF_EXCLAVES) $(DOCONF_SPTM) $(DOCONF_KERNEL_TAG) $(CURRENT_KERNEL_CONFIG)
+	$(_v)$(SRCROOT)/SETUP/config/doconf -c -cpu $(DOCONF_ARCH_CONFIG_LC) -soc $(CURRENT_MACHINE_CONFIG_LC) -platform $(PLATFORM_FOR_CONFIG) -d $(TARGET)/$(CURRENT_KERNEL_CONFIG) -s $(SOURCE) -m $(MASTERCONFDIR) $(DOCONF_SEED) $(DOCONF_EXCLAVES) $(DOCONF_SPTM) $(DOCONF_KERNEL_TAG) $(DOCONF_OSSHW) $(CURRENT_KERNEL_CONFIG)
 
 do_all: $(TARGET)/$(CURRENT_KERNEL_CONFIG)/Makefile
 	$(_v)${MAKE} \
diff --git a/libsyscall/Libsyscall.xcconfig b/libsyscall/Libsyscall.xcconfig
index 8b4b2197..3137114e 100644
--- a/libsyscall/Libsyscall.xcconfig
+++ b/libsyscall/Libsyscall.xcconfig
@@ -1,4 +1,4 @@
-#include "<DEVELOPER_DIR>/Makefiles/CoreOS/Xcode/BSD.xcconfig"
+
 
 BUILD_VARIANTS = normal
 SUPPORTED_PLATFORMS = macosx iphoneos iphoneosnano tvos appletvos watchos bridgeos driverkit
diff --git a/makedefs/MakeInc.def b/makedefs/MakeInc.def
index e37762bc..1971214a 100644
--- a/makedefs/MakeInc.def
+++ b/makedefs/MakeInc.def
@@ -29,10 +29,11 @@ SUPPORTED_ARM64_MACHINE_CONFIGS = NONE
 else ifneq ($(findstring _host,$(RC_ProjectName)),)
 SUPPORTED_ARM64_MACHINE_CONFIGS = NONE
 else
-SUPPORTED_ARM64_MACHINE_CONFIGS = T6000 T6020 T6030 T6031 T6041 T6050 T8101 T8103 T8112 T8122 T8132 T8142 VMAPPLE
+SUPPORTED_ARM64_MACHINE_CONFIGS = T6000 T6020 T6030 T6031 T6041 T6050 T8101 T8103 T8112 T8122 T8132 T8142 VMAPPLE QEMU SUPERBIRD IPAD41
 
 SPTM_ENABLED_SOCS_MacOSX = t6000 t8112 t6020 t8122 t6030 t6031 t6041 t6050 t8132 t8142
 EXCLAVES_ENABLED_SOCS_MacOSX = t8142 t6050
+OSSHW_SOCS = qemu superbird ipad41
 
 
 endif
@@ -94,6 +95,10 @@ MACHINE_FLAGS_ARM64_T8142 = -DARM64_BOARD_CONFIG_T8142 			-mcpu=apple-m4+memtag+
 
 MACHINE_FLAGS_ARM64_VMAPPLE = -DARM64_BOARD_CONFIG_VMAPPLE              -march=armv8.5a+sme2
 
+MACHINE_FLAGS_ARM64_QEMU = -DARM64_BOARD_CONFIG_QEMU -mcpu=cortex-a53
+MACHINE_FLAGS_ARM64_SUPERBIRD = -DARM64_BOARD_CONFIG_SUPERBIRD -mcpu=cortex-a53
+MACHINE_FLAGS_ARM64_IPAD41 = -DARM64_BOARD_CONFIG_IPAD41 -mcpu=cyclone
+
 
 #
 # Deployment target flag
@@ -354,6 +359,10 @@ ARCH_STRING_FOR_CURRENT_MACHINE_CONFIG ?=
 
 ifeq ($(ARCH_STRING_FOR_CURRENT_MACHINE_CONFIG),)
 
+ifneq ($(filter qemu superbird ipad41,$(CURRENT_MACHINE_CONFIG_LC)),)
+export ARCH_STRING_FOR_CURRENT_MACHINE_CONFIG := arm64
+endif
+
 ifneq ($(EMBEDDED_DEVICE_MAP),)
 export ARCH_STRING_FOR_CURRENT_MACHINE_CONFIG := $(shell $(EMBEDDED_DEVICE_MAP) -db $(EDM_DBPATH) -list -query SELECT DISTINCT KernelMachOArchitecture FROM Targets WHERE KernelPlatform IS \"$(CURRENT_MACHINE_CONFIG_LC)\" LIMIT 1 || echo UNKNOWN )
 ifeq ($(ARCH_STRING_FOR_CURRENT_MACHINE_CONFIG),)
@@ -1151,7 +1160,8 @@ LD_KERNEL_LIBS	   = -lcc_kext
 LD_KERNEL_ARCHIVES = $(LDFLAGS_KERNEL_SDK) -lfirehose_kernel
 
 # Link binary support library
-ifneq ($(CURRENT_MACHINE_CONFIG),NONE)
+ifneq ($(filter $(CURRENT_MACHINE_CONFIG),NONE $(shell echo $(OSSHW_SOCS) | tr a-z A-Z)),)
+else
         LDFLAGS_KERNEL_ONLY += -rdynamic -Wl,-force_load,$(KDKROOT)/System/Library/KernelSupport/lib$(CURRENT_MACHINE_CONFIG).os.$(CURRENT_KERNEL_CONFIG).a
 endif
 
@@ -1179,6 +1189,10 @@ ifneq ($(filter $(CURRENT_MACHINE_CONFIG_LC),$(EXCLAVES_ENABLED_SOCS_$(PLATFORM)
   EXCLAVES_ENABLED := 1
 endif
 
+ifneq ($(filter $(CURRENT_MACHINE_CONFIG_LC),$(OSSHW_SOCS)),)
+  DOCONF_OSSHW := -osshw
+endif
+
 ifeq ($(SPTM_ENABLED),1)
   LD_KERNEL_ARCHIVES += -lTrustedExecutionMonitor_Kernel
   LD_KERNEL_ARCHIVES += -lsptm_xnu
diff --git a/osfmk/conf/Makefile b/osfmk/conf/Makefile
index d6dceb52..dc33bfd9 100644
--- a/osfmk/conf/Makefile
+++ b/osfmk/conf/Makefile
@@ -23,7 +23,7 @@ endif
 
 $(TARGET)/$(CURRENT_KERNEL_CONFIG)/Makefile: $(SRCROOT)/SETUP/config/doconf $(OBJROOT)/SETUP/config $(DOCONFDEPS)
 	$(_v)$(MKDIR) $(TARGET)/$(CURRENT_KERNEL_CONFIG)
-	$(_v)$(SRCROOT)/SETUP/config/doconf -c -cpu $(DOCONF_ARCH_CONFIG_LC) -soc $(CURRENT_MACHINE_CONFIG_LC) -platform $(PLATFORM_FOR_CONFIG) -d $(TARGET)/$(CURRENT_KERNEL_CONFIG) -s $(SOURCE) -m $(MASTERCONFDIR) $(DOCONF_SEED) $(DOCONF_EXCLAVES) $(DOCONF_SPTM) $(DOCONF_KERNEL_TAG) $(CURRENT_KERNEL_CONFIG)
+	$(_v)$(SRCROOT)/SETUP/config/doconf -c -cpu $(DOCONF_ARCH_CONFIG_LC) -soc $(CURRENT_MACHINE_CONFIG_LC) -platform $(PLATFORM_FOR_CONFIG) -d $(TARGET)/$(CURRENT_KERNEL_CONFIG) -s $(SOURCE) -m $(MASTERCONFDIR) $(DOCONF_SEED) $(DOCONF_EXCLAVES) $(DOCONF_SPTM) $(DOCONF_KERNEL_TAG) $(DOCONF_OSSHW) $(CURRENT_KERNEL_CONFIG)
 
 do_all: $(TARGET)/$(CURRENT_KERNEL_CONFIG)/Makefile
 	$(_v)${MAKE} \
diff --git a/pexpert/conf/Makefile b/pexpert/conf/Makefile
index d6dceb52..dc33bfd9 100644
--- a/pexpert/conf/Makefile
+++ b/pexpert/conf/Makefile
@@ -23,7 +23,7 @@ endif
 
 $(TARGET)/$(CURRENT_KERNEL_CONFIG)/Makefile: $(SRCROOT)/SETUP/config/doconf $(OBJROOT)/SETUP/config $(DOCONFDEPS)
 	$(_v)$(MKDIR) $(TARGET)/$(CURRENT_KERNEL_CONFIG)
-	$(_v)$(SRCROOT)/SETUP/config/doconf -c -cpu $(DOCONF_ARCH_CONFIG_LC) -soc $(CURRENT_MACHINE_CONFIG_LC) -platform $(PLATFORM_FOR_CONFIG) -d $(TARGET)/$(CURRENT_KERNEL_CONFIG) -s $(SOURCE) -m $(MASTERCONFDIR) $(DOCONF_SEED) $(DOCONF_EXCLAVES) $(DOCONF_SPTM) $(DOCONF_KERNEL_TAG) $(CURRENT_KERNEL_CONFIG)
+	$(_v)$(SRCROOT)/SETUP/config/doconf -c -cpu $(DOCONF_ARCH_CONFIG_LC) -soc $(CURRENT_MACHINE_CONFIG_LC) -platform $(PLATFORM_FOR_CONFIG) -d $(TARGET)/$(CURRENT_KERNEL_CONFIG) -s $(SOURCE) -m $(MASTERCONFDIR) $(DOCONF_SEED) $(DOCONF_EXCLAVES) $(DOCONF_SPTM) $(DOCONF_KERNEL_TAG) $(DOCONF_OSSHW) $(CURRENT_KERNEL_CONFIG)
 
 do_all: $(TARGET)/$(CURRENT_KERNEL_CONFIG)/Makefile
 	$(_v)${MAKE} \
diff --git a/san/conf/Makefile b/san/conf/Makefile
index d6dceb52..dc33bfd9 100644
--- a/san/conf/Makefile
+++ b/san/conf/Makefile
@@ -23,7 +23,7 @@ endif
 
 $(TARGET)/$(CURRENT_KERNEL_CONFIG)/Makefile: $(SRCROOT)/SETUP/config/doconf $(OBJROOT)/SETUP/config $(DOCONFDEPS)
 	$(_v)$(MKDIR) $(TARGET)/$(CURRENT_KERNEL_CONFIG)
-	$(_v)$(SRCROOT)/SETUP/config/doconf -c -cpu $(DOCONF_ARCH_CONFIG_LC) -soc $(CURRENT_MACHINE_CONFIG_LC) -platform $(PLATFORM_FOR_CONFIG) -d $(TARGET)/$(CURRENT_KERNEL_CONFIG) -s $(SOURCE) -m $(MASTERCONFDIR) $(DOCONF_SEED) $(DOCONF_EXCLAVES) $(DOCONF_SPTM) $(DOCONF_KERNEL_TAG) $(CURRENT_KERNEL_CONFIG)
+	$(_v)$(SRCROOT)/SETUP/config/doconf -c -cpu $(DOCONF_ARCH_CONFIG_LC) -soc $(CURRENT_MACHINE_CONFIG_LC) -platform $(PLATFORM_FOR_CONFIG) -d $(TARGET)/$(CURRENT_KERNEL_CONFIG) -s $(SOURCE) -m $(MASTERCONFDIR) $(DOCONF_SEED) $(DOCONF_EXCLAVES) $(DOCONF_SPTM) $(DOCONF_KERNEL_TAG) $(DOCONF_OSSHW) $(CURRENT_KERNEL_CONFIG)
 
 do_all: $(TARGET)/$(CURRENT_KERNEL_CONFIG)/Makefile
 	$(_v)${MAKE} \
diff --git a/security/conf/Makefile b/security/conf/Makefile
index d6dceb52..dc33bfd9 100644
--- a/security/conf/Makefile
+++ b/security/conf/Makefile
@@ -23,7 +23,7 @@ endif
 
 $(TARGET)/$(CURRENT_KERNEL_CONFIG)/Makefile: $(SRCROOT)/SETUP/config/doconf $(OBJROOT)/SETUP/config $(DOCONFDEPS)
 	$(_v)$(MKDIR) $(TARGET)/$(CURRENT_KERNEL_CONFIG)
-	$(_v)$(SRCROOT)/SETUP/config/doconf -c -cpu $(DOCONF_ARCH_CONFIG_LC) -soc $(CURRENT_MACHINE_CONFIG_LC) -platform $(PLATFORM_FOR_CONFIG) -d $(TARGET)/$(CURRENT_KERNEL_CONFIG) -s $(SOURCE) -m $(MASTERCONFDIR) $(DOCONF_SEED) $(DOCONF_EXCLAVES) $(DOCONF_SPTM) $(DOCONF_KERNEL_TAG) $(CURRENT_KERNEL_CONFIG)
+	$(_v)$(SRCROOT)/SETUP/config/doconf -c -cpu $(DOCONF_ARCH_CONFIG_LC) -soc $(CURRENT_MACHINE_CONFIG_LC) -platform $(PLATFORM_FOR_CONFIG) -d $(TARGET)/$(CURRENT_KERNEL_CONFIG) -s $(SOURCE) -m $(MASTERCONFDIR) $(DOCONF_SEED) $(DOCONF_EXCLAVES) $(DOCONF_SPTM) $(DOCONF_KERNEL_TAG) $(DOCONF_OSSHW) $(CURRENT_KERNEL_CONFIG)
 
 do_all: $(TARGET)/$(CURRENT_KERNEL_CONFIG)/Makefile
 	$(_v)${MAKE} \
